src/libmongoc/src/mongoc/mongoc-uri.c

```diff
@@ -1746,6 +1768,72 @@ mongoc_uri_finalize_auth (mongoc_uri_t *uri, bson_error_t *error)
       // Defer remaining validation of `MongoCredential` fields to Authentication Handshake.
    }
 
+   // MONGODB-OIDC
+   else if (strcasecmp (mechanism, "MONGODB-OIDC") == 0) {
+      // Authentication spec: username: MAY be specified (with callback/environment defined meaning).
+      if (!_finalize_auth_username (username, mechanism, _mongoc_uri_finalize_allowed, error)) {
+         goto fail;
+      }
+
+      // Authentication spec: source: MUST be "$external" and defaults to "$external".
+      if (!_finalize_auth_source_default_external (uri, source, mechanism, error)) {
+         goto fail;
+      }
+
+      // Authentication spec: password: MUST NOT be specified.
+      if (!_finalize_auth_password (password, mechanism, _mongoc_uri_finalize_prohibited, error)) {
+         goto fail;
+      }
+
+      // mechanism_properties are allowed for MONGODB-OIDC.
+      if (!_finalize_auth_oidc_mechanism_properties (mechanism_properties, error)) {
+         goto fail;
+      }
+
+      // The environment is optional, but if specified it must appear valid.
+      if (mechanism_properties && bson_iter_init_find_case (&iter, mechanism_properties, "ENVIRONMENT")) {
+         const mongoc_oidc_env_t *env = mongoc_oidc_env_find (bson_iter_utf8 (&iter, NULL));
+         if (!env) {
```

Comment:
```suggestion
         if (!BSON_ITER_HOLDS_UTF8 (&iter)) {
            MONGOC_URI_ERROR (...);
            goto fail;
         }

         if (!env) {
```

`bson_iter_utf8` may return `NULL` if a user sets the `credentials` or `authentication_mechanism` field incorrectly (not a string). This is fine for `mongoc_oidc_env_find` but not for `printf` formatting below (in error message).